# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/shared/CodeGenerator-shared.cpp
# Commit: 1c35c1a7b373
# Full Hash: 1c35c1a7b3731ecbc70be235430a6f06376494d1
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2022-06-09 22:00:48
# Regressor Bug: 1770366
# File Overlap Count: 2
# Description:
#   Bug 1770366 part 5 - Add frame pointer to IonJS/Warp frames. r=iain
#   
#   Change OSR to use the standard frame epilogue on the Baseline side and frame prologue
#   on the CodeGenerator side.
#   
# ==============================================================================

diff -r e8f787c59185 -r 1c35c1a7b373 js/src/jit/shared/CodeGenerator-shared.cpp
--- a/js/src/jit/shared/CodeGenerator-shared.cpp	Thu Jun 09 13:49:16 2022 +0000
+++ b/js/src/jit/shared/CodeGenerator-shared.cpp	Thu Jun 09 13:49:17 2022 +0000
@@ -102,12 +102,16 @@
                                           WasmStackAlignment);
     }
   } else {
-    // Round to JitStackAlignment, and implicitly to sizeof(Value) as
-    // JitStackAlignment is a multiple of sizeof(Value). This was originally
-    // implemented for SIMD.js, but now lets us use faster ABI calls via
-    // setupAlignedABICall.
-    MOZ_ASSERT(offsetOfLocalSlots_ == 0);
-    frameDepth_ = AlignBytes(graph->localSlotsSize(), JitStackAlignment);
+    // Reserve space for frame pointer (and padding on 32-bit platforms).
+    offsetOfLocalSlots_ = JitFrameLayout::IonFirstSlotOffset;
+    frameDepth_ = offsetOfLocalSlots_;
+
+    // Allocate space for local slots (register allocator spills). Round to
+    // JitStackAlignment, and implicitly to sizeof(Value) as JitStackAlignment
+    // is a multiple of sizeof(Value). This was originally implemented for
+    // SIMD.js, but now lets us use faster ABI calls via setupAlignedABICall.
+    frameDepth_ += graph->localSlotsSize();
+    frameDepth_ = AlignBytes(frameDepth_, JitStackAlignment);
 
     // Allocate space for argument Values passed to callee functions.
     offsetOfPassedArgSlots_ = frameDepth_;
@@ -134,8 +138,13 @@
   // Ensure that the Ion frame is properly aligned.
   masm.assertStackAlignment(JitStackAlignment, 0);
 
+  // Frame prologue.
+  masm.Push(FramePointer);
+  masm.moveStackPtrTo(FramePointer);
+
   // Note that this automatically sets MacroAssembler::framePushed().
-  masm.reserveStack(frameSize());
+  masm.reserveStack(frameSize() - sizeof(uintptr_t));
+  MOZ_ASSERT(masm.framePushed() == frameSize());
   masm.checkStackAlignment();
 
   if (JS::TraceLoggerSupported()) {
@@ -153,8 +162,10 @@
     emitTracelogIonStop();
   }
 
-  masm.freeStack(frameSize());
-  MOZ_ASSERT(masm.framePushed() == 0);
+  MOZ_ASSERT(masm.framePushed() == frameSize());
+  masm.moveToStackPtr(FramePointer);
+  masm.pop(FramePointer);
+  masm.setFramePushed(0);
 
   // If profiling, reset the per-thread global lastJitFrame to point to
   // the previous frame.
@@ -336,7 +347,7 @@
 static inline int32_t ToStackIndex(LAllocation* a) {
   if (a->isStackSlot()) {
     MOZ_ASSERT(a->toStackSlot()->slot() >= 1);
-    return a->toStackSlot()->slot();
+    return JitFrameLayout::IonFirstSlotOffset + a->toStackSlot()->slot();
   }
   return -int32_t(sizeof(JitFrameLayout) + a->toArgument()->index());
 }
